File: protocol/csharp/ToInfinity.Protocol.Tests/InputEventTests.cs
using ToInfinity.Protocol;
using Xunit;

namespace ToInfinity.Protocol.Tests;

public class InputEventTests
{
    [Fact]
    public void MouseMove_RoundTrips()
    {
        var evt = new MouseMoveEvent(0, 0);
        byte[] encoded = evt.Encode();
        Assert.Equal(5, encoded.Length);

        var decoded = Assert.IsType<MouseMoveEvent>(InputEvent.TryDecode(encoded));
        Assert.Equal(0, decoded.X);
        Assert.Equal(0, decoded.Y);
    }

    [Fact]
    public void MouseMove_MaxValues_RoundTrips()
    {
        var evt = new MouseMoveEvent(ushort.MaxValue, ushort.MaxValue);
        var decoded = Assert.IsType<MouseMoveEvent>(InputEvent.TryDecode(evt.Encode()));
        Assert.Equal(ushort.MaxValue, decoded.X);
        Assert.Equal(ushort.MaxValue, decoded.Y);
    }

    [Theory]
    [InlineData(MouseButton.Left)]
    [InlineData(MouseButton.Right)]
    [InlineData(MouseButton.Middle)]
    [InlineData(MouseButton.X1)]
    [InlineData(MouseButton.X2)]
    public void MouseButtonDown_RoundTrips(MouseButton button)
    {
        var evt = new MouseButtonDownEvent(100, 200, button);
        byte[] encoded = evt.Encode();
        Assert.Equal(6, encoded.Length);

        var decoded = Assert.IsType<MouseButtonDownEvent>(InputEvent.TryDecode(encoded));
        Assert.Equal(100, decoded.X);
        Assert.Equal(200, decoded.Y);
        Assert.Equal(button, decoded.Button);
    }

    [Fact]
    public void MouseButtonUp_RoundTrips()
    {
        var evt = new MouseButtonUpEvent(100, 200, MouseButton.Right);
        byte[] encoded = evt.Encode();
        Assert.Equal(6, encoded.Length);

        var decoded = Assert.IsType<MouseButtonUpEvent>(InputEvent.TryDecode(encoded));
        Assert.Equal(100, decoded.X);
        Assert.Equal(200, decoded.Y);
        Assert.Equal(MouseButton.Right, decoded.Button);
    }

    [Fact]
    public void MouseWheel_RoundTrips_PositiveAndNegativeDeltas()
    {
        var evt = new MouseWheelEvent(1, 2, -120, 120);
        byte[] encoded = evt.Encode();
        Assert.Equal(9, encoded.Length);

        var decoded = Assert.IsType<MouseWheelEvent>(InputEvent.TryDecode(encoded));
        Assert.Equal(1, decoded.X);
        Assert.Equal(2, decoded.Y);
        Assert.Equal(-120, decoded.DeltaX);
        Assert.Equal(120, decoded.DeltaY);
    }

    [Fact]
    public void MouseWheel_MinMaxDeltas_RoundTrip()
    {
        var evt = new MouseWheelEvent(0, 0, short.MinValue, short.MaxValue);
        var decoded = Assert.IsType<MouseWheelEvent>(InputEvent.TryDecode(evt.Encode()));
        Assert.Equal(short.MinValue, decoded.DeltaX);
        Assert.Equal(short.MaxValue, decoded.DeltaY);
    }

    [Fact]
    public void KeyDown_RoundTrips()
    {
        // HID Usage 0x04 = keyboard "A" on a US layout.
        var evt = new KeyDownEvent(0x04);
        byte[] encoded = evt.Encode();
        Assert.Equal(3, encoded.Length);

        var decoded = Assert.IsType<KeyDownEvent>(InputEvent.TryDecode(encoded));
        Assert.Equal(0x04, decoded.HidUsage);
    }

    [Fact]
    public void KeyUp_RoundTrips_MaxHidUsage()
    {
        var evt = new KeyUpEvent(ushort.MaxValue);
        var decoded = Assert.IsType<KeyUpEvent>(InputEvent.TryDecode(evt.Encode()));
        Assert.Equal(ushort.MaxValue, decoded.HidUsage);
    }

    [Fact]
    public void EmptyPayload_ReturnsNull()
    {
        Assert.Null(InputEvent.TryDecode(ReadOnlySpan<byte>.Empty));
    }

    [Fact]
    public void UnknownEventKind_ReturnsNull()
    {
        byte[] payload = { 0xFF, 0x00, 0x00 };
        Assert.Null(InputEvent.TryDecode(payload));
    }

    [Theory]
    [InlineData(InputEventKind.MouseMove, 4)] // needs 5
    [InlineData(InputEventKind.MouseButtonDown, 5)] // needs 6
    [InlineData(InputEventKind.MouseWheel, 8)] // needs 9
    [InlineData(InputEventKind.KeyDown, 2)] // needs 3
    public void TruncatedPayload_ReturnsNull(InputEventKind kind, int shortLength)
    {
        var payload = new byte[shortLength];
        payload[0] = (byte)kind;
        Assert.Null(InputEvent.TryDecode(payload));
    }
}

File: protocol/swift/ToInfinityProtocol/Tests/ToInfinityProtocolTests/InputEventTests.swift
import XCTest
@testable import ToInfinityProtocol

final class InputEventTests: XCTestCase {
    func testMouseMoveRoundTrips() {
        let event = InputEvent.mouseMove(x: 0, y: 0)
        let encoded = event.encode()
        XCTAssertEqual(encoded.count, 5)

        guard case let .mouseMove(x, y)? = InputEvent.decode(from: encoded) else {
            XCTFail("expected .mouseMove")
            return
        }
        XCTAssertEqual(x, 0)
        XCTAssertEqual(y, 0)
    }

    func testMouseMoveMaxValuesRoundTrips() {
        let event = InputEvent.mouseMove(x: .max, y: .max)
        guard case let .mouseMove(x, y)? = InputEvent.decode(from: event.encode()) else {
            XCTFail("expected .mouseMove")
            return
        }
        XCTAssertEqual(x, UInt16.max)
        XCTAssertEqual(y, UInt16.max)
    }

    func testMouseButtonDownRoundTrips() {
        for button: MouseButton in [.left, .right, .middle, .x1, .x2] {
            let event = InputEvent.mouseButtonDown(x: 100, y: 200, button: button)
            let encoded = event.encode()
            XCTAssertEqual(encoded.count, 6)

            guard case let .mouseButtonDown(x, y, decodedButton)? = InputEvent.decode(from: encoded) else {
                XCTFail("expected .mouseButtonDown")
                return
            }
            XCTAssertEqual(x, 100)
            XCTAssertEqual(y, 200)
            XCTAssertEqual(decodedButton, button)
        }
    }

    func testMouseButtonUpRoundTrips() {
        let event = InputEvent.mouseButtonUp(x: 100, y: 200, button: .right)
        let encoded = event.encode()
        XCTAssertEqual(encoded.count, 6)

        guard case let .mouseButtonUp(x, y, button)? = InputEvent.decode(from: encoded) else {
            XCTFail("expected .mouseButtonUp")
            return
        }
        XCTAssertEqual(x, 100)
        XCTAssertEqual(y, 200)
        XCTAssertEqual(button, .right)
    }

    func testMouseWheelRoundTripsPositiveAndNegativeDeltas() {
        let event = InputEvent.mouseWheel(x: 1, y: 2, deltaX: -120, deltaY: 120)
        let encoded = event.encode()
        XCTAssertEqual(encoded.count, 9)

        guard case let .mouseWheel(x, y, deltaX, deltaY)? = InputEvent.decode(from: encoded) else {
            XCTFail("expected .mouseWheel")
            return
        }
        XCTAssertEqual(x, 1)
        XCTAssertEqual(y, 2)
        XCTAssertEqual(deltaX, -120)
        XCTAssertEqual(deltaY, 120)
    }

    func testMouseWheelMinMaxDeltasRoundTrip() {
        let event = InputEvent.mouseWheel(x: 0, y: 0, deltaX: .min, deltaY: .max)
        guard case let .mouseWheel(_, _, deltaX, deltaY)? = InputEvent.decode(from: event.encode()) else {
            XCTFail("expected .mouseWheel")
            return
        }
        XCTAssertEqual(deltaX, Int16.min)
        XCTAssertEqual(deltaY, Int16.max)
    }

    func testKeyDownRoundTrips() {
        // HID Usage 0x04 = keyboard "A" on a US layout.
        let event = InputEvent.keyDown(hidUsage: 0x04)
        let encoded = event.encode()
        XCTAssertEqual(encoded.count, 3)

        guard case let .keyDown(hidUsage)? = InputEvent.decode(from: encoded) else {
            XCTFail("expected .keyDown")
            return
        }
        XCTAssertEqual(hidUsage, 0x04)
    }

    func testKeyUpRoundTripsMaxHidUsage() {
        let event = InputEvent.keyUp(hidUsage: .max)
        guard case let .keyUp(hidUsage)? = InputEvent.decode(from: event.encode()) else {
            XCTFail("expected .keyUp")
            return
        }
        XCTAssertEqual(hidUsage, UInt16.max)
    }

    func testEmptyPayloadReturnsNil() {
        XCTAssertNil(InputEvent.decode(from: Data()))
    }

    func testUnknownEventKindReturnsNil() {
        let payload = Data([0xFF, 0x00, 0x00])
        XCTAssertNil(InputEvent.decode(from: payload))
    }

    func testTruncatedPayloadReturnsNil() {
        let cases: [(InputEventKind, Int)] = [
            (.mouseMove, 4), // needs 5
            (.mouseButtonDown, 5), // needs 6
            (.mouseWheel, 8), // needs 9
            (.keyDown, 2), // needs 3
        ]

        for (kind, shortLength) in cases {
            var payload = [UInt8](repeating: 0, count: shortLength)
            payload[0] = kind.rawValue
            XCTAssertNil(InputEvent.decode(from: Data(payload)), "kind \(kind) with length \(shortLength) should fail to decode")
        }
    }
}
